src/prolly/leaf.ts

```diff
@@ -0,0 +1,139 @@
+      builder,
+      LeafFB.createEntriesVector(builder, leafEntries),
+    );
+    builder.finish(root);
+    const data = builder.asUint8Array();
+
+    const chunk = Chunk.new(data, []);
+    return new Leaf(chunk);
+  }
+
+  static load(chunk: Chunk): Leaf {
+    // Validate at load-time so we can assume data is valid thereafter.
+    const buf = new flatbuffers.ByteBuffer(chunk.data);
+    const root = LeafFB.getRootAsLeaf(buf);
+
+    // Rust has a 'missing entries' error but the TS/JS bindings has no way to
+    // distinguish if the entries is present and empty vs missing.
+
+    const entriesLength = root.entriesLength();
+
+    let prev: Uint8Array | null = null;
+    for (let i = 0; i < entriesLength; i++) {
+      const entry = root.entries(i);
+      assertNotNull(entry);
+      const ek = entry.keyArray();
+
+      if (prev !== null) {
+        if (!ek) {
+          throw new Error('missing key');
+        }
+        const ord = arrayCompare(prev, ek);
+        if (ord === 0) {
+          throw new Error('duplicate key');
+        }
+        if (ord > 0) {
+          throw new Error('unsorted key');
+        }
+      }
+      if (!ek) {
+        throw new Error('missing key');
+      }
+      if (entry.valArray() === null) {
+        throw new Error('missing val');
+      }
+      prev = ek;
+    }
+
+    return new Leaf(chunk);
+  }
+
+  *entries(): Generator<Entry, void> {
+    const buf = new flatbuffers.ByteBuffer(this.chunk.data);
+    const root = LeafFB.getRootAsLeaf(buf);
+    for (let i = 0; i < root.entriesLength(); i++) {
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      const entry = root.entries(i)!;
+      yield {
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        key: entry.keyArray()!,
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        val: entry.valArray()!,
+      };
+    }
+  }
+
+  [Symbol.iterator](): Generator<Entry, void> {
+    return this.entries();
+  }
+
+  getEntryByIndex(i: number): LeafEntryFB {
+    const buf = new flatbuffers.ByteBuffer(this.chunk.data);
+    const root = LeafFB.getRootAsLeaf(buf);
+    const entry = root.entries(i);
+    assertNotNull(entry);
+    return entry;
+  }
+
+  binarySearch(key: Uint8Array): {found: boolean; index: number} {
+    const buf = new flatbuffers.ByteBuffer(this.chunk.data);
+    const root = LeafFB.getRootAsLeaf(buf);
+
+    let size = root.entriesLength();
+    if (size === 0) {
+      return {found: false, index: 0};
+    }
+    let base = 0;
+
+    while (size > 1) {
+      const half = Math.floor(size / 2);
+      const mid = base + half;
+      // mid is always in [0, size), that means mid is >= 0 and < size.
+      // mid >= 0: by definition
+      // mid < size: mid = size / 2 + size / 4 + size / 8 ...
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      const entry = root.entries(mid)!;
```

Comment:
This is going to have to iterate the fb from zero each time it is called!

Reply:
This is a case where the complexity of fb is obscuring things and hiding problems. I'm not sure it's a net win. Looking forward to experimenting with JSON.

Reply:
That is not accurate. `entries(idx)` is O(1)

Reply:
You're right, I was misreading this.